runtime/src/accounts_db.rs

```diff
@@ -629,6 +629,42 @@ impl AccountsDB {
         false
     }
 
+    // Reclaim older states of rooted non-zero lamports accounts as a general
+    // AccountsDB bloat mitigation and preprocess for better zero-lamport purging.
+    fn compact_before_purge_zero_lamport_accounts(
+        &self,
+        purges: &HashMap<Pubkey, Vec<(Slot, AccountInfo)>>,
+    ) {
+        let mut accounts_index = self.accounts_index.write().unwrap();
+        let mut all_pubkeys: HashSet<Pubkey> = HashSet::<Pubkey>::default();
+        for root in accounts_index.not_compacted_roots.iter() {
+            let pubkey_sets: Vec<HashSet<Pubkey>> = self.scan_account_storage(
+                *root,
+                |stored_account: &StoredAccount,
+                 _store_id: AppendVecId,
+                 accum: &mut HashSet<Pubkey>| {
+                    accum.insert(stored_account.meta.pubkey);
+                },
+            );
+
+            for pubkey_set in pubkey_sets {
+                for pubkey in pubkey_set {
+                    all_pubkeys.insert(pubkey);
+                }
+            }
+        }
+        let mut reclaims = Vec::new();
+        for pubkey in all_pubkeys {
```

Comment:
I haven't measured how much this impacts the overall validator performance yet...

Reply:
Basically, this incurs slot_vec write locks for the union of all updated accounts since last checked. Effectively this means one more such lock for all account updates.

Reply:
Yea. it would be nice to see how it affects it. It seems the main cost is the accounts scan though, so this might not be too much. I dont have a good idea of how many accounts would be purged this way. So maybe we should test it in a testnet.